File: modules/student-experience/src/main/java/com/campusos/portal/service/AttendanceMath.java
package com.campusos.portal.service;

import com.campusos.portal.domain.AttendanceRecord;
import java.util.List;

public final class AttendanceMath {

    private AttendanceMath() {}

    /** PRESENT + APPROVED_LEAVE over every day that counts. SCHEDULED days are excluded. */
    public static Double pct(List<AttendanceRecord> rows) {
        long counted = rows.stream().filter(AttendanceRecord::counted).count();
        if (counted == 0) return null;
        long attended = rows.stream().filter(AttendanceRecord::counted)
                .filter(AttendanceRecord::attended).count();
        return Math.round(attended * 1000.0 / counted) / 10.0;
    }
}

File: modules/student-experience-js/src/service/AttendanceMath.js
export const AttendanceMath = {
  /** PRESENT + APPROVED_LEAVE over every day that counts. SCHEDULED days are excluded. */
  pct(rows) {
    const counted = rows.filter((r) => r.status !== 'SCHEDULED');
    if (counted.length === 0) return null;
    const attended = counted.filter((r) => r.status === 'PRESENT' || r.status === 'APPROVED_LEAVE');
    return Math.round((attended.length * 1000) / counted.length) / 10;
  },
};
